src/v/serde/test/serde_test.cc

```diff
@@ -1077,3 +1079,16 @@ SEASTAR_THREAD_TEST_CASE(collections_interop) {
     BOOST_REQUIRE(serialized_vector == serialized_fifo);
     BOOST_REQUIRE(serialized_vector == serialized_f_vector);
 }
+
+SEASTAR_THREAD_TEST_CASE(variant) {
+    using my_variant = std::variant<ss::sstring, int, bool, my_enum, int>;
```

Comment:
this seems problematic because it's not apparent how one could version a change to the variant. it's generally the case in serde that the primitives are unversioned because their encoding is forever. that's the case here, too, right?

Reply:
WRT to compatibility: you can append to the options in a variant and as long as old versions don't read that it's fine (so the feature table dance to wait for the cluster to upgrade then you can start writing the new variant).

In terms of versioning, it's encouraged that you wrap these in a struct/envelope for versioning, which is what I do in my PR. I'll expand on these details in the commit and add a comment to the file.

Reply:
Added a comment and updated the cover letter